### DISCUSSION TITLE
Actions upon death :)

### DISCUSSION THREAD

**Discussion by FrozenEmpire** [ChilliHugger/The-Lords-Of-Midnight#173] [Code Algorithm Questions] [answered]
Hi Chris,

Could you please make a couple of additional code checks / clarifications when you get chance 
This was triggered by considering how many columns I need on the new "Loyalty" screen in The Icemark :)

### DDR - Death of a liege
I was curious what happened to the vassals when their liege would die (either during a Fight, or Battle)
ie Do the vassals then have no liege, does their liege become some higher up commander, or does their liege remain "as is" even though they are dead

In ddr_character::whatIsCharacterDoing, in the cases of OD_FOLLOW_LIEGE and OD_FOLLOW_FOE, there is a test that if the followed character is dead, then try to reassign the liege of the dead liege to the vassal, or if that was nobody, then make it default to Luxor.

Q: That's only for AI Controlled characters. What happens for recruited vassals whose liege dies somehow ?
Q: For AI Controlled characters. What happens if they have other orders than above and their liege dies ?

I couldn't find any handling for those cases, so assume the liege remains "as is" even though they died. Is that right ?

### LOM & DDR - Death of a group leader / follower
Similar curiosity to above, but what happens to group leaders / followers upon death (again during a Fight, or Battle)

In ddr_battle::battleArmies, I can see that if a defending group leader / follower loses a battle then they either disband the group, or leave the group, before the dead check is made.

Q: The above is only for DDR. I couldn't see any equivalent for LOM death in battle other than the exception in mxscenario::CheckMoonringWearerDead
Q: What happens if a follower dies through a Fight ? Do they have to be manually removed from the group ?
Q: What happens if a group leader dies through a Fight ? Does the group have to be manually disbanded ?

Regards,
Andrew

**Comment by IcemarkUK** [comment, marked as answer]
Hi Andrew..

### LOM & DDR - Death of a group leader / follower

As grouping is a feature that I added there is a good chance I may have missed some scenarios.

I can confirm that 
- in LoM groups are not disbanded when the follower dies in a battle.
- in both the group is not disbanded when a lord dies in a fight.
- in both a follower is not removed when they die in any way. It is unlikely they die as a fight because the leader usually bears the brunt of this, but they could die in a battle.

I've raised bugs for these issues for me. But it did raise an interesting point for me. Currently the UI would keep the dead lord and the user could at least see that the leader or follower was dead and disband or remove the lord themselves. Even swapping out a leader if they wanted to without disbanding the group. Although that feature doesn't seem to work when the leader is dead! :)

Removing the lord from the group - there is a chance that the user would not notice the dead lord!

There is a part of me that thinks that disbanding the group automatically in DDR is the wrong thing... I'm going to contemplate this and maybe that is the bug fix that will add.

<img width="491" alt="Screenshot 2023-08-31 at 12 22 45" src="https://github.com/ChilliHugger/The-Lords-Of-Midnight/assets/8301416/6a71b349-1c22-491d-879b-d0dc987ea2a6">
<img width="734" alt="Screenshot 2023-08-31 at 12 29 44" src="https://github.com/ChilliHugger/The-Lords-Of-Midnight/assets/8301416/9019799e-63f5-48f7-989e-015497496738">

**Reply by FrozenEmpire** [reply]
Thanks for clarifying Chris,

I quite like the fact that you can see a group leader/member has died and have to decide how to regroup because of that.

the only issue I can see with removing the auto-ungroup when losing in battle is that it seems a necessary pre-requisite for displacement.

I guess for displacement they could ungroup as now, but if they actually died then they could remain grouped and leave the player to decide (as for fights)

**Reply by IcemarkUK** [reply]
I wonder if displacement is needed when grouping. It only gives single lords a chance to escape. I'll check the code to make sure that it isn't to stop them being assessed for future hits in the same night battle. If that is the case they could easily be removed from the battle but stay in place.

**Reply by FrozenEmpire** [reply]
yeah that would work for grouped specific behaviour. Single lords can get displaced as per original.

I guess safety in numbers is enough to perhaps preventing them fleeing :)

**Reply by IcemarkUK** [reply]
I'm thinking of attaching it to Difficulty. On most modes always keep the group intact. But in difficult then allow the lords to be displaced. That means the player has to pay more attention to their groups.

**Reply by FrozenEmpire** [reply]
Sounds cool

**Comment by IcemarkUK** [comment]
### DDR - Death of a liege

#### Q: That's only for AI Controlled characters. What happens for recruited vassals whose liege dies somehow ?
This is never updated. I checked the code and played the game a little to make sure.
From an updated POV I think the lord should pick the liege of their dead liege. But from a difficulty POV I might be inclined to make than random or desert the cause - based on stats or flip of a coin.

#### Q: For AI Controlled characters. What happens if they have other orders than above and their liege dies ?
The AI character will eventually pick a new liege when/if their orders get reassigned to OD_FOLLOW_LIEGE. There is nothing in the original code to handle this scenario other than that,


<img width="1297" alt="Screenshot 2023-08-31 at 12 58 51" src="https://github.com/ChilliHugger/The-Lords-Of-Midnight/assets/8301416/e2180d62-fcfb-42f3-8819-cc4299f3b9cb">
<img width="1299" alt="Screenshot 2023-08-31 at 12 59 32" src="https://github.com/ChilliHugger/The-Lords-Of-Midnight/assets/8301416/530d7232-de33-4f72-9ce9-248e46145f98">

**Reply by FrozenEmpire** [reply]
Thanks also on this point

Reassigning to the liege of the dead liege seems good to me - in both non-AI and AI cases, irrespective of current orders.

However apart from then not having a dead liege, I’m not sure what the implications of doing nothing actually would be. Do you see any influence on gameplay by reassigning liege upon death? I still need to deepen my understanding of the logic :)

**Reply by IcemarkUK** [reply]
I'll read over the finer points, but I don't think it will make any difference to game play. It's one really there to help with the orders - and potentially recruitment - ie: When an AI lord is recruited by their liege there is a bonus +3 to help the outcome.

**Reply by FrozenEmpire** [reply]
If recruitment is affected, then maybe to stay true to original it should be left as is, with the dead liege remaining.

Then the other possibilities only for extra difficulty options, which will make for slightly adjusted recruitment behaviour both AI and non AI

**Comment by FrozenEmpire** [comment]
An additional point to clarify within DDR scenario upon death

If a recruited character dies whilst in a tunnel (eg Fighting), then I believe they drop any object they are carrying...
... into the location above them ?

(Also a related point - did I understand right that any object that can be fought at a given location, will appear both above ground and in the tunnel at that location ?)

**Comment by IcemarkUK** [comment]
> If a recruited character dies whilst in a tunnel (eg Fighting), then I believe they drop any object they are carrying... ... into the location above them ?

It's not actively done, but that is true. Those objects have no concept of the tunnels, so they are always above ground when not carried. ie: A lord has an in tunnel flag, but the object does not.

> (Also a related point - did I understand right that any object that can be fought at a given location, will appear both above ground and in the tunnel at that location ?)

No that's no correct. Basically where a tunnel runs under plains, mountains, hills, and forest the critter is below ground. Everywhere else the critter is above ground.


```
mxthing_t mxmap::getLocationObject( const mxcharacter* c, mxgridref loc )
{
    mxloc& l = GetAt ( loc );
    
#if defined(_LOM_)
    return (mxthing_t)l.object;
#endif
    
#if defined(_DDR_)
    
    if ( !(l.flags&lf_creature) )
        return OB_NONE ;
    
    //
    if ( l.IsTunnelPassageway() ) {
        // location is a passageway
        if ( c->IsInTunnel() )
            return (mxthing_t)l.object ;
    }else{
        if ( !c->IsInTunnel() ) {
            if ( (l.terrain != TN_ICYWASTE) && (l.terrain != TN_FROZENWASTE) )
                return (mxthing_t)l.object ;
        }
        
    }
#endif
    return OB_NONE;
}

```

```
bool mxloc::IsTunnelPassageway() const
{
    mxterrain_t t = mx->scenario->toScenarioTerrain((mxterrain_t)terrain);
    return HasTunnel() && (t ==TN_PLAINS2 || t==TN_MOUNTAIN2 || t==TN_FOREST2 || t==TN_HILLS ) ;
}

```

**Reply by FrozenEmpire** [reply]
Got it. Subtle, but when 48k is the limit you need to shoe horn it in somehow :) Many thanks

**Reply by FrozenEmpire** [reply]
It occurred to me that there is an obvious opportunity as a difficulty option to track things and unique objects accurately above or below ground. So some stuff is placed in tunnels on scenario creation, characters pickup/drop from above or below ground, etc.

so if a character carrying one of the unique objects dies in a tunnel, the unique object can only be picked up from inside the tunnel etc

one for the ideas list :)

**Reply by IcemarkUK** [reply]
Yes, I have this in mind for The Citadel as it has more underground tunnels/dungeons. My plan was to just add a of_tunnel that works the same was as cf_tunnel.

**Comment by FrozenEmpire** [comment]
Bring it on ;)
